dali/pipeline/operators/op_spec_test.cc

```diff
@@ -0,0 +1,169 @@
+    for (int i = 0; i < out0.shape().num_samples(); i++) {
+      *out0.mutable_tensor<int>(i) = i;
+    }
+
+    auto &out1 = ws.OutputRef<CPUBackend>(1);
+    for (int i = 0; i < out1.shape()[0][0]; i++) {
+      out1.mutable_data<float>()[i] = i;
+    }
+
+    auto &out2 = ws.OutputRef<CPUBackend>(2);
+    for (int i = 0; i < out2.shape().num_samples(); i++) {
+      *out2.mutable_tensor<int>(i) = i;
+    }
+
+    auto &out3 = ws.OutputRef<CPUBackend>(3);
+    for (int i = 0; i < out3.shape().num_samples(); i++) {
+      for (int j = 0; j < 2; j++) {
+        out3.mutable_tensor<int>(i)[j] = i;
+      }
+    }
+  }
+};
+
+DALI_REGISTER_OPERATOR(SupportGeneratingOp, SupportGeneratingOp, Support);
+
+DALI_SCHEMA(SupportGeneratingOp)
+  .DocStr("SupportGeneratingOp")
+  .NumInput(0)
+  .NumOutput(4);
+
+
+class SupportAccessingOp : public Operator<CPUBackend> {
+ public:
+  explicit SupportAccessingOp(const OpSpec &spec) : Operator<CPUBackend>(spec) {}
+
+  bool CanInferOutputs() const override {
+    return true;
+  }
+
+  bool SetupImpl(std::vector<OutputDesc> &output_desc, const HostWorkspace &ws) override {
+    output_desc.resize(1);
+    output_desc[0] = {kernels::uniform_list_shape(batch_size_, {1}), TypeInfo::Create<int>()};
+    return true;
+  }
+
+  void RunImpl(HostWorkspace &ws) override {
+    for (int i = 0; i < batch_size_; i++) {
+      EXPECT_EQ(spec_.GetArgument<int>("arg0", &ws, i), i);
+      EXPECT_EQ(spec_.GetArgument<float>("arg1", &ws, i), i);
+    }
+    // Non-matching shapes (differnet than 1 scalar value per sample) should not work with
+    // OpSpec::GetArgument()
+    ASSERT_THROW(auto z = spec_.GetArgument<float>("arg2", &ws, 0), std::runtime_error);
+    ASSERT_THROW(auto z = spec_.GetArgument<float>("arg3", &ws, 0), std::runtime_error);
+
+    // They can be accessed as proper ArgumentInputs
+    auto &ref_2 = ws.ArgumentInput("arg2");
+    ASSERT_EQ(ref_2.shape().num_samples(), batch_size_ + 5);
+    ASSERT_TRUE(kernels::is_uniform(ref_2.shape()));
+    ASSERT_EQ(ref_2.shape()[0], kernels::TensorShape<>(1));
+    for (int i = 0; i < ref_2.shape().num_samples(); i++) {
+      EXPECT_EQ(ref_2.tensor<int>(i)[0], i);
+    }
+
+    auto &ref_3 = ws.ArgumentInput("arg3");
+    ASSERT_EQ(ref_3.shape().num_samples(), batch_size_);
+    ASSERT_TRUE(kernels::is_uniform(ref_3.shape()));
+    ASSERT_EQ(ref_3.shape()[0], kernels::TensorShape<>(1, 2));
+    for (int i = 0; i < ref_3.shape().num_samples(); i++) {
+      for (int j = 0; j < 2; j++) {
+        EXPECT_EQ(ref_3.tensor<int>(i)[j], i);
+      }
+    }
+  }
+};
+
+DALI_REGISTER_OPERATOR(SupportAccessingOp, SupportAccessingOp, CPU);
+
+DALI_SCHEMA(SupportAccessingOp)
+  .DocStr("SupportAccessingOp")
+  .NumInput(0)
+  .NumOutput(1)
+  .AddOptionalArg("arg0", "no-doc", 42, true)
+  .AddOptionalArg("arg1", "no-doc", 42.f, true)
+  .AddOptionalArg("arg2", "no-doc", 42, true)
+  .AddOptionalArg("arg3", "no-doc", 42, true);
+
+TEST(ArgumentInputTest, OpSpecAccess) {
```

Comment:
Very cool way to test this :) Maybe add a bit of comment as it is not "standard" test.

Reply:
Done